### PR TITLE
Refactoring of monitoring service and infra in order to allow getting metrics by type and by domain

### PR DESCRIPTION
**Pull request by eugene-cheverda**
Refactoring of monitoring service and infra in order to allow getting metrics by type and by domain only (getting latter to work is TBD). Also these changes are preparation for managing different http errors (badrequest, notfound, etc)

### CODE REVIEW

File: core/src/main/scala/kafka/console/core/providers/monitoring/MBeanInfoProvider.scala [hunk 1]
```diff
@@ -25,6 +25,11 @@ object MBeanInfoProvider {
       .map(a => a._1 + "=" + a._2).mkString(",")
   }
 
+  private def getMetricType(a: ObjectName) = canonicalNameKeyListToMap(a.getCanonicalKeyPropertyListString)
+    .find(_._1 == "type")
+    .map(_._2)
```
Comment (theocat) [inline core/src/main/scala/kafka/console/core/providers/monitoring/MBeanInfoProvider.scala]: `case (_, value) => value`

File: service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala [hunk 2]
```diff
@@ -5,43 +5,65 @@ package app
...
+      definePotentialMetricsCandidates(domain, mtype, keys).map(l => l.map(b => convertSingle(b, domain, b.mtype, mbcs, attributes)))
+    ).attempt
+  }
+
+  private def definePotentialMetricsCandidates(domain: String, mtype: Option[String], keys: Map[String, Option[String]]): Task[List[MBeanMetricInfo]] = {
+    mtype match {
+      case None => aggregates.map(_.filter(a => s"${a.domain}" == domain).toList)
+      case Some(t) => aggregates.map(_.filter(a =>
+        if (keys.isEmpty) s"${a.domain}" == domain && s"${a.mtype}" == t
+        else s"${a.domain}" == domain && s"${a.mtype}" == t && s"${a.canonicalName}" == buildCanonicalKeysList(mtype, keys)).toList
```
Comment (theocat) [inline service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala]: it just should be like that... consider http://eed3si9n.com/learning-scalaz/Equal.html instead

File: service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala [hunk 3]
```diff
@@ -5,43 +5,65 @@ package app
...
-        .toArray
-        .map(_.asInstanceOf[Attribute])
-        .map(a => MetricAttribute(a.getName, a.getValue.toString))
-        .toList
-    }.map(a => List(KafkaMetric(domain, mtype, a))).getOrElse(Nil)
+  private def buildCanonicalKeysList(mtype: Option[String], keys: Map[String, Option[String]]) = {
+    (keys + ("type" -> mtype))
+      .toList.filter(a => a._2.isDefined)
+      .sortBy(a => a._1)
+      .map(a => a._1 + "=" + a._2.get)
```
Comment (theocat) [inline service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala]: same

File: service/src/main/scala/kafka/console/service/Application.scala [hunk 4]
```diff
@@ -36,18 +36,31 @@ object Application {
...
-    case GET -> Root / "monitoring" / domain / mtype :? NameMatcher(name) +& ClientIdMatcher(clientId) +& AttributesMatcher(attributes) => {
-      val attrs = attributes match {
+    case GET -> Root / "monitoring" / domain / mtype :? params => {
+      val attributes = params.get("attributes").map(s => s.head) match {
         case Some(a) => a.split(",").toList
         case _ => Nil
       }
+      val keys = List("name" -> params.get("name").map(s => s.head),
+        "client-id" -> params.get("clientId").map(s => s.head)
+      ).filter(_._2.isDefined).toMap
```
Comment (theocat) [inline service/src/main/scala/kafka/console/service/Application.scala]: same

Comment (eugene-cheverda) [inline service/src/main/scala/kafka/console/service/Application.scala]: Fixed

File: service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala (hunk 3 above)
Comment (eugene-cheverda) [inline service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala]: Fixed

File: service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala (hunk 2 above)
Comment (eugene-cheverda) [inline service/src/main/scala/kafka/console/service/app/AppMonitoringService.scala]: Fixed

File: core/src/main/scala/kafka/console/core/providers/monitoring/MBeanInfoProvider.scala (hunk 1 above)
Comment (eugene-cheverda) [inline core/src/main/scala/kafka/console/core/providers/monitoring/MBeanInfoProvider.scala]: Fixed

**Pull request merged**